File: cli/Sources/botspeaker/ControlClient.swift
import Foundation

/// Locates the running BotSpeaker app through its discovery file and talks
/// to its loopback control API.
struct ControlClient {
    struct Discovery: Decodable {
        let url: String
        let port: Int
        let token: String
        let pid: Int
        let version: String
    }

    struct Failure: Error, CustomStringConvertible {
        let exitCode: Int32
        let code: String
        let message: String
        var description: String { message }
    }

    let baseURL: URL
    let token: String
    let discovery: Discovery?

    static let bundleID = "ai.DJBen.2.BotSpeaker"

    static var discoveryCandidates: [URL] {
        let home = FileManager.default.homeDirectoryForCurrentUser
        var urls: [URL] = []
        if let override = ProcessInfo.processInfo.environment["BOTSPEAKER_CONTROL_FILE"], !override.isEmpty {
            urls.append(URL(fileURLWithPath: override))
        }
        urls.append(home
            .appendingPathComponent("Library/Containers/\(bundleID)/Data/Library/Application Support/BotSpeaker/control.json"))
        urls.append(home.appendingPathComponent("Library/Application Support/BotSpeaker/control.json"))
        return urls
    }

    /// Returns a client for a live app instance, or nil when no discovery
    /// file points at a running process.
    static func findRunning() -> (client: ControlClient?, stalePID: Int?) {
        var stalePID: Int?
        for candidate in discoveryCandidates {
            guard let data = try? Data(contentsOf: candidate),
                  let discovery = try? JSONDecoder().decode(Discovery.self, from: data),
                  let base = URL(string: discovery.url) else { continue }
            if kill(pid_t(discovery.pid), 0) != 0 && errno == ESRCH {
                // Stale file from a previous launch.
                stalePID = discovery.pid
                continue
            }
            return (ControlClient(baseURL: base, token: discovery.token, discovery: discovery), nil)
        }
        return (nil, stalePID)
    }

    /// Finds the running app, launching it in the background first when it is
    /// not running. Set `BOTSPEAKER_NO_LAUNCH=1` to fail instead of launching,
    /// and `BOTSPEAKER_APP=/path/to/BotSpeaker.app` to launch a specific copy.
    static func locate(launchIfNeeded: Bool = true) async throws -> ControlClient {
        let environment = ProcessInfo.processInfo.environment
        if let url = environment["BOTSPEAKER_CONTROL_URL"], let token = environment["BOTSPEAKER_TOKEN"],
           let base = URL(string: url) {
            return ControlClient(baseURL: base, token: token, discovery: nil)
        }
        var found = findRunning()
        if found.client == nil, launchIfNeeded, environment["BOTSPEAKER_NO_LAUNCH"].map({ $0.isEmpty || $0 == "0" }) ?? true {
            try launchApp(explicitPath: environment["BOTSPEAKER_APP"])
            let deadline = Date().addingTimeInterval(launchTimeout)
            while found.client == nil, Date() < deadline {
                try await Task.sleep(nanoseconds: 250_000_000)
                found = findRunning()
            }
            if found.client == nil {
                throw Failure(
                    exitCode: 2,
                    code: "app_unreachable",
                    message: "Launched BotSpeaker but it did not publish a control file within \(Int(launchTimeout)) seconds. "
                        + "Check that the installed app is version 0.4 or newer (the one built from this repository), "
                        + "or point BOTSPEAKER_APP at the right BotSpeaker.app."
                )
            }
        }
        if let client = found.client { return client }
        let detail = found.stalePID.map { "the last control.json belongs to pid \($0), which has exited" }
            ?? "no control.json found"
        throw Failure(
            exitCode: 2,
            code: "app_unreachable",
            message: "BotSpeaker is not running (\(detail)). Launch the BotSpeaker app, then retry. Looked in: "
                + discoveryCandidates.map(\.path).joined(separator: ", ")
        )
    }

    static let launchTimeout: TimeInterval = 20

    /// Asks LaunchServices to start the app without stealing focus.
    private static func launchApp(explicitPath: String?) throws {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/open")
        if let explicitPath, !explicitPath.isEmpty {
            process.arguments = ["-g", explicitPath]
        } else {
            process.arguments = ["-g", "-b", bundleID]
        }
        let stderr = Pipe()
        process.standardOutput = FileHandle.nullDevice
        process.standardError = stderr
        do {
            try process.run()
        } catch {
            throw Failure(exitCode: 2, code: "app_launch_failed", message: "Could not run /usr/bin/open: \(error.localizedDescription)")
        }
        process.waitUntilExit()
        if process.terminationStatus != 0 {
            let output = String(data: stderr.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            throw Failure(
                exitCode: 2,
                code: "app_launch_failed",
                message: "BotSpeaker is not installed or could not be launched (\(output.isEmpty ? "open exited \(process.terminationStatus)" : output)). "
                    + "Install BotSpeaker.app or set BOTSPEAKER_APP to its path."
            )
        }
    }

    func get(_ path: String, query: [String: String] = [:]) async throws -> [String: Any] {
        try await send(method: "GET", path: path, query: query, body: nil)
    }

    func post(_ path: String, body: [String: Any] = [:]) async throws -> [String: Any] {
        try await send(method: "POST", path: path, query: [:], body: body)
    }

    private func send(method: String, path: String, query: [String: String], body: [String: Any]?) async throws -> [String: Any] {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if !query.isEmpty {
            components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        var request = URLRequest(url: components.url!)
        request.httpMethod = method
        request.timeoutInterval = 3600
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw Failure(
                exitCode: 2,
                code: "app_unreachable",
                message: "Could not reach BotSpeaker at \(baseURL.absoluteString): \(error.localizedDescription). Is the app running?"
            )
        }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        let payload = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        if status == 401 {
            throw Failure(exitCode: 2, code: "unauthorized", message: "BotSpeaker rejected the control token. Restart the app or clear BOTSPEAKER_TOKEN.")
        }
        if status >= 400 || payload["ok"] as? Bool == false, let error = payload["error"] as? [String: Any] {
            throw Failure(
                exitCode: 1,
                code: error["code"] as? String ?? "error",
                message: error["message"] as? String ?? "Request failed with HTTP \(status)."
            )
        }
        if status >= 400 {
            throw Failure(exitCode: 1, code: "http_\(status)", message: "Request failed with HTTP \(status).")
        }
        return payload
    }
}

File: Windows/BotSpeakerCli/ControlClient.cs
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace BotSpeaker.Cli;

/// <summary>
/// Locates the running BotSpeaker app through its discovery file and talks to
/// its loopback control API — the Windows counterpart of the macOS CLI's
/// ControlClient.
/// </summary>
public sealed class ControlClient
{
    public sealed record Discovery(string Url, int Port, string Token, int Pid, string Version, string? Exe);

    /// <summary>Where the CLI remembers the last app it talked to, so it can relaunch it.</summary>
    private static string MemoryPath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "BotSpeaker", "cli.json");

    public sealed class Failure(int exitCode, string code, string message) : Exception(message)
    {
        public int ExitCode { get; } = exitCode;
        public string Code { get; } = code;
    }

    public Uri BaseUrl { get; }
    public string Token { get; }
    public Discovery? Found { get; }

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromHours(1) };
    private static readonly TimeSpan LaunchTimeout = TimeSpan.FromSeconds(20);

    private ControlClient(Uri baseUrl, string token, Discovery? found)
    {
        BaseUrl = baseUrl;
        Token = token;
        Found = found;
    }

    public static IEnumerable<string> DiscoveryCandidates
    {
        get
        {
            var overridePath = Environment.GetEnvironmentVariable("BOTSPEAKER_CONTROL_FILE");
            if (!string.IsNullOrEmpty(overridePath)) yield return overridePath;
            yield return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "BotSpeaker", "control.json");
        }
    }

    /// <summary>Returns a client for a live app instance, or null (plus any stale pid seen).</summary>
    public static (ControlClient? Client, int? StalePid) FindRunning()
    {
        int? stalePid = null;
        foreach (var candidate in DiscoveryCandidates)
        {
            Discovery? discovery;
            try
            {
                if (!File.Exists(candidate)) continue;
                var node = JsonNode.Parse(File.ReadAllText(candidate)) as JsonObject;
                if (node is null) continue;
                discovery = new Discovery(
                    node["url"]?.GetValue<string>() ?? "",
                    node["port"]?.GetValue<int>() ?? 0,
                    node["token"]?.GetValue<string>() ?? "",
                    node["pid"]?.GetValue<int>() ?? 0,
                    node["version"]?.GetValue<string>() ?? "",
                    node["exe"]?.GetValue<string>());
            }
            catch (Exception)
            {
                continue;
            }
            if (discovery.Url.Length == 0 || discovery.Token.Length == 0) continue;
            if (!IsProcessAlive(discovery.Pid))
            {
                stalePid = discovery.Pid;
                continue;
            }
            RememberApp(discovery.Exe);
            return (new ControlClient(new Uri(discovery.Url), discovery.Token, discovery), null);
        }
        return (null, stalePid);
    }

    private static void RememberApp(string? exe)
    {
        if (string.IsNullOrEmpty(exe) || RememberedApp() == exe) return;
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(MemoryPath)!);
            File.WriteAllText(MemoryPath, new JsonObject { ["lastApp"] = exe }.ToJsonString());
        }
        catch (Exception)
        {
            // Purely a convenience; launching falls back to the fixed candidates.
        }
    }

    private static string? RememberedApp()
    {
        try
        {
            if (!File.Exists(MemoryPath)) return null;
            return (JsonNode.Parse(File.ReadAllText(MemoryPath)) as JsonObject)?["lastApp"]?.GetValue<string>();
        }
        catch (Exception)
        {
            return null;
        }
    }

    private static bool IsProcessAlive(int pid)
    {
        if (pid <= 0) return false;
        try
        {
            using var process = Process.GetProcessById(pid);
            return !process.HasExited;
        }
        catch (ArgumentException)
        {
            return false;
        }
        catch (InvalidOperationException)
        {
            return false;
        }
    }

    /// <summary>
    /// Finds the running app, launching it in the background first when it is
    /// not running. Set BOTSPEAKER_NO_LAUNCH=1 to fail instead of launching,
    /// and BOTSPEAKER_APP=C:\path\to\BotSpeaker.exe to launch a specific copy.
    /// </summary>
    public static async Task<ControlClient> LocateAsync()
    {
        var url = Environment.GetEnvironmentVariable("BOTSPEAKER_CONTROL_URL");
        var token = Environment.GetEnvironmentVariable("BOTSPEAKER_TOKEN");
        if (!string.IsNullOrEmpty(url) && !string.IsNullOrEmpty(token))
        {
            return new ControlClient(new Uri(url), token, null);
        }
        var found = FindRunning();
        var noLaunch = Environment.GetEnvironmentVariable("BOTSPEAKER_NO_LAUNCH");
        bool mayLaunch = string.IsNullOrEmpty(noLaunch) || noLaunch == "0";
        if (found.Client is null && mayLaunch)
        {
            LaunchApp();
            var deadline = DateTime.UtcNow + LaunchTimeout;
            while (found.Client is null && DateTime.UtcNow < deadline)
            {
                await Task.Delay(250);
                found = FindRunning();
            }
            if (found.Client is null)
            {
                throw new Failure(2, "app_unreachable",
                    $"Launched BotSpeaker but it did not publish a control file within {(int)LaunchTimeout.TotalSeconds} seconds. "
                    + "Check that the installed app is 0.4.1 or newer (a build that has the control API), or point BOTSPEAKER_APP at the right BotSpeaker.exe.");
            }
        }
        if (found.Client is ControlClient client)
        {
            client.WarnIfAppIsNewer();
            return client;
        }
        var detail = found.StalePid is int stale
            ? $"the last control.json belongs to pid {stale}, which has exited"
            : "no control.json found";
        throw new Failure(2, "app_unreachable",
            $"BotSpeaker is not running ({detail}). Launch the BotSpeaker app, then retry. Looked in: {string.Join(", ", DiscoveryCandidates)}");
    }

    /// <summary>
    /// Candidate app locations, most specific first: BOTSPEAKER_APP, the app
    /// this CLI last talked to, a BotSpeaker.exe beside this CLI, the standard
    /// install location, and a Desktop copy (how the portable zip is usually
    /// run).
    /// </summary>
    public static IEnumerable<string> AppCandidates
    {
        get
        {
            var explicitPath = Environment.GetEnvironmentVariable("BOTSPEAKER_APP");
            if (!string.IsNullOrEmpty(explicitPath)) yield return explicitPath;
            if (RememberedApp() is string remembered) yield return remembered;
            yield return Path.Combine(AppContext.BaseDirectory, "BotSpeaker.exe");
            yield return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Programs", "BotSpeaker", "BotSpeaker.exe");
            yield return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory), "BotSpeaker.exe");
        }
    }

    private static void LaunchApp()
    {
        var exe = AppCandidates.FirstOrDefault(File.Exists)
            ?? throw new Failure(2, "app_launch_failed",
                "BotSpeaker.exe was not found. Start the BotSpeaker app once by hand (the CLI remembers where it is), "
                + "put it beside this CLI, or set BOTSPEAKER_APP to its path. Looked in: "
                + string.Join(", ", AppCandidates));
        try
        {
            var info = new ProcessStartInfo(exe, "--background")
            {
                UseShellExecute = true,
                WorkingDirectory = Path.GetDirectoryName(exe) ?? "",
            };
            Process.Start(info);
        }
        catch (Exception error)
        {
            throw new Failure(2, "app_launch_failed", $"Could not launch {exe}: {error.Message}");
        }
    }

    /// <summary>
    /// Nudges the user to upgrade when the app is newer than this binary.
    /// Printed to stderr so JSON output on stdout stays clean. Set
    /// BOTSPEAKER_NO_UPGRADE_HINT=1 to silence it.
    /// </summary>
    private void WarnIfAppIsNewer()
    {
        if (Found is null || Environment.GetEnvironmentVariable("BOTSPEAKER_NO_UPGRADE_HINT") is not null) return;
        if (Version.TryParse(Found.Version, out var app) && Version.TryParse(CliVersion.Current, out var cli) && cli < app)
        {
            Console.Error.WriteLine($"note: BotSpeaker app is {Found.Version} but this CLI is {CliVersion.Current}. Rebuild or reinstall the CLI from the matching release.");
        }
    }

    public Task<JsonObject> GetAsync(string path, Dictionary<string, string>? query = null) =>
        SendAsync(HttpMethod.Get, path, query, null);

    public Task<JsonObject> PostAsync(string path, JsonObject? body = null) =>
        SendAsync(HttpMethod.Post, path, null, body ?? []);

    private async Task<JsonObject> SendAsync(HttpMethod method, string path, Dictionary<string, string>? query, JsonObject? body)
    {
        var builder = new UriBuilder(BaseUrl) { Path = path };
        if (query is { Count: > 0 })
        {
            builder.Query = string.Join("&", query.Select(pair => $"{Uri.EscapeDataString(pair.Key)}={Uri.EscapeDataString(pair.Value)}"));
        }
        using var request = new HttpRequestMessage(method, builder.Uri);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", Token);
        if (body is not null)
        {
            request.Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
        }

        HttpResponseMessage response;
        string text;
        try
        {
            response = await Http.SendAsync(request);
            text = await response.Content.ReadAsStringAsync();
        }
        catch (Exception error) when (error is HttpRequestException or TaskCanceledException)
        {
            throw new Failure(2, "app_unreachable", $"Could not reach BotSpeaker at {BaseUrl}: {error.Message} Is the app running?");
        }
        int status = (int)response.StatusCode;
        JsonObject payload;
        try { payload = JsonNode.Parse(text) as JsonObject ?? []; }
        catch (JsonException) { payload = []; }
        if (status == 401)
        {
            throw new Failure(2, "unauthorized", "BotSpeaker rejected the control token. Restart the app or clear BOTSPEAKER_TOKEN.");
        }
        bool notOk = payload["ok"] is JsonValue okValue && okValue.TryGetValue<bool>(out var ok) && !ok;
        if ((status >= 400 || notOk) && payload["error"] is JsonObject errorBody)
        {
            throw new Failure(1,
                errorBody["code"]?.GetValue<string>() ?? "error",
                errorBody["message"]?.GetValue<string>() ?? $"Request failed with HTTP {status}.");
        }
        if (status >= 400)
        {
            throw new Failure(1, $"http_{status}", $"Request failed with HTTP {status}.");
        }
        return payload;
    }
}

public static class CliVersion
{
    public static string Current =>
        typeof(CliVersion).Assembly.GetCustomAttributes(typeof(System.Reflection.AssemblyInformationalVersionAttribute), false)
            .OfType<System.Reflection.AssemblyInformationalVersionAttribute>()
            .FirstOrDefault()?.InformationalVersion.Split('+')[0]
        ?? typeof(CliVersion).Assembly.GetName().Version?.ToString(3)
        ?? "0.0.0";
}
